src/MongoDB.Driver.Core/Core/Bindings/CoreSession.cs

```diff
@@ -347,6 +356,7 @@ public void Dispose()
         public void StartTransaction(TransactionOptions transactionOptions = null)
         {
             EnsureStartTransactionCanBeCalled();
+            _pinnedShardRouter = null;
```

Comment:
I've noted in XML documentation for {{PinnedShardRouter}} that "Value has meaning if and only if a sharded transaction is in progress."

Reply:
You could unpin on abort, but I don't think you should unpin on commit, because applications are allowed to call commit multiple times (weird, but true)

Reply:
While working on this for Go driver, Isabella tried to unpin but ran into issues with the tests from https://github.com/mongodb/specifications/commit/08c94e6d6b16416faaa336208d696087d7cc7b7c, which is the same version the .NET driver is synced against. Therefore, I recommend deferring unpinning until the implementation of recovery tokens